Adding recommendation for connecting multiple coprocessors (#195)
openhwgroup/core-v-xif: docs/source/x_ext.rst

For the purpose of relative identification, an instruction is considered to be preceding another instruction, if it was accepted in an issue transaction at an earlier time.
The other instruction is thus succeeding the earlier one.

Multiple |coprocessors|
~~~~~~~~~~~~~~~~~~~~~~~

This specification defines a point-to-point connection between a |processor| and a |coprocessor|, that is defined in a way that facilitates the integration of multiple |coprocessors|.

> The combined interface of the |coprocessors| must adhere to this specification.

Reviewer: ```suggestion
The combined interface of the |coprocessors| must adhere to this specification and thus must behave like a single |coprocessor| from the |processor| point of view.
```

  +------------------------+----------------------+-----------------------------------------------------------------------------------------------------------------+
  | Signal                 | Type                 | Description                                                                                                     |
  +========================+======================+=================================================================================================================+

>   | ``instr``              | logic [31:0]         | Uncompressed instruction. Must be signalled as 0 if ``accept`` is 0.                                            |

Author: similar statements are already in place for all fields of the issue_resp. I think it would be good to have it consistent. Might be a topic to discuss next wednesday. I will forward you the invite to the call

>   | ``instr``              | logic [31:0]         | Uncompressed instruction. Must be signalled as 0 if ``accept`` is 0.                                            |

Reviewer: Sorry I didn't catch this on ``issue_resp`` during my initial review.
Just a general comment from my side:
- OK for consistent behaviour on all interface
- policy enforcement should be avoided in a specification if alternative solutions may exist to solve the same problem.

My understanding here is that the proposed policy is linked to a particular solution (OR-combining response signal) when connecting multiple coprocessors to the same interface. Since a bit of glue logic is required to do that, it is up to the glue logic to solve the problem.

>   | ``instr``              | logic [31:0]         | Uncompressed instruction. Must be signalled as 0 if ``accept`` is 0.                                            |

Reviewer 2: I don't fully agree that there is a consistency requirement w.r.t. forcing signals to 0. In the issue interface for example writeback = 1 is defined as 'accepted plus writeback required'. I would be against forcing all kind of signals to 0 for consistency. If you really want consistency then I would rather remove the forcing to 0 on the issue interface (and other interfaces if it exists).

have exactly one result transaction (even if no data needs to be written back to the |processor|'s register file). No result transaction shall be performed for instructions which have not been accepted for offload or
for instructions that have been killed.

> If ``result_valid`` is 0, all signals within ``result`` must also be 0.

Reviewer: Same comment as above.
```suggestion
If ``result_valid`` is 0, all signals within ``result`` must be ignored.
```
Forcing them to 0 is too restrictive and may prevent power optimizations. Simply ignoring them opens the door to more design choices.
In the multiple coprocessor scenario, it is up to the signal combination logic to deal with this, based on the choice that was made by each coprocessor provider.

Author: point taken. i would like to get this discussed in a wider audience.

  * Clearly document the supported and required interfaces.

Recommendations for implementing multiple coprocessors on a shared interface
----------------------------------------------------------------------------

It is possible to implement multiple |coprocessors|, which connect to a single |processor|.
There is no required implementation to do this, but the specification facilitates the integration with several measures.

> This section provides details per interface.

Reviewer: ```suggestion
This section provides detailed guidelines per interface for a simple use case where all |coprocessors| have exactly the same timing on their interfaces.
More complex recombination logic may be required when connecting |coprocessors| that have different timing specifications on interface signals.
```

* Compressed interface

  The ``compressed_valid`` and ``compressed_req`` signals can be broadcasted to all |coprocessors|.
  Each |coprocessor| drives its ``compressed_ready`` and ``compressed_resp`` signals.
  It is recommended that |coprocessors| provide the response within the same cycle.
  In this case, both will be driving ``compressed_ready`` the same way.

>   It is possible to form an OR combination of the ``compressed_resp`` signals, as a coprocessor rejecting the instruction will signal all zeros.

Reviewer: ```suggestion
  It is possible to form an OR combination of the ``compressed_resp`` signals, by clamping all of them to zeros (either inside the |coprocessor| or in the signals recombination logic).
```

  More than one |coprocessor| accepting an instruction must be prevented by design. 

* Issue interface

  The ``issue_valid`` and ``issue_req`` signals can be broadcasted to all |coprocessors|.
  Each |coprocessor| drives its ``issue_ready`` and ``issue_resp`` signals.
  It is recommended that |coprocessors| provide the response within the same cycle.
  In this case, both will be driving ``issue_ready`` the same way.
  It is possible to form an OR combination of the ``issue_resp`` signals, as a coprocessor rejecting the instruction will signal all zeros.
  More than one |coprocessor| accepting an instruction must be prevented by design. 

* Register interface

  The ``register_valid`` and ``register_req`` signals can be broadcasted to all |coprocessors|.
  Each |coprocessor| drives its ``register_ready`` signal.
  The transition of a coprocessors ``register_ready`` signal to 1 should only occur when it is clear that the next transaction is targeting an instruction accepted by it.
  This can be the case, if there is a clear sequence from issue to register interface, or when ``register_valid`` = 1 and ``id`` is matching an instruction accepted by the |coprocessor|.

>   It is possible to provide an OR combination of the ``register_ready`` signals as a combined signal to the |processor|.

Author: this is a single signal, so I think this comment does not quite apply here.

The de-multiplexing logic can be reduced to a simple OR combination, if the output signals of the |coprocessors| not mapped to the instruction are 0.
This applies to the compressed interface, the issue interface, and the result interface.

* Compressed interface

  The ``compressed_valid`` and ``compressed_req`` signals can be broadcasted to all |coprocessors|.
  Each |coprocessor| drives its ``compressed_ready`` and ``compressed_resp`` signals.
  It is recommended that |coprocessors| provide the response within the same cycle.
  In this case, both will be driving ``compressed_ready`` the same way.
  The ``compressed_resp`` signals need to be de-multiplexed based on the ``compressed_resp.accept`` signals. 
  More than one |coprocessor| accepting an instruction must be prevented by design. 

* Issue interface

  The ``issue_valid`` and ``issue_req`` signals can be broadcasted to all |coprocessors|.
  Each |coprocessor| drives its ``issue_ready`` and ``issue_resp`` signals.
  It is recommended that |coprocessors| provide the response within the same cycle.
  In this case, both will be driving ``issue_ready`` the same way.

>   The ``compressed_resp`` signals need to be de-multiplexed based on the ``issue_resp.accept`` signals. 

Reviewer: ```suggestion
  The ``issue_resp`` signals need to be de-multiplexed based on the ``issue_resp.accept`` signals. 
```

  This can be the case, if there is a clear sequence from issue to register interface, or when ``register_valid`` = 1 and ``id`` is matching an instruction accepted by the |coprocessor|.
  It is possible to provide an OR combination of the ``register_ready`` signals as a combined signal to the |processor|.

* Commit interface

  The commit transaction is unidirectional from the |processor| to the |coprocessor|.
  All signals will be broadcasted to all |coprocessors|.
  The definition of the commit interface requires the |coprocessors| to be functional if faced with ``id`` values it did not accept.

.. only:: MemoryIf

  * Memory (request/response) interface

  At time of writing, there are no recommendation for the Memory (request/response) interface yet.

  * Memory result interface

  At time of writing, there are no recommendation for the Memory result interface interface yet.

* Result interface

  The ``result_ready`` signal can be broadcasted to all |coprocessors|.
  Each |coprocessor| drives its ``result_valid`` signal.

>   If all instructions in all |coprocessors| complete execution in a fixed number of |processor| clock cycles after their register interface transaction completed, and writeback is never stalled (i.e. ``result_ready`` is 1 at that time), it is possible to provide de-multiplex the ``result`` based on the ``result_valid`` signals.

Reviewer: ```suggestion
  If all instructions in all |coprocessors| complete execution in a fixed number of |processor| clock cycles after their register interface transaction completed, and writeback is never stalled (i.e. ``result_ready`` is 1 at that time), it is possible to de-multiplex the ``result`` based on the ``result_valid`` signals.
```
